File: iosApp/iosApp/Modules/CurrencyReceive ✅/CurrencyReceiveWireframe.swift
// Created by web3d4v on 13/06/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import UIKit
import web3walletcore

final class DefaultCurrencyReceiveWireframe {
    private weak var parent: UIViewController?
    private let context: CurrencyReceiveWireframeContext
    private let networksService: NetworksService
    
    private weak var vc: UIViewController?
    
    init(
        _ parent: UIViewController?,
        context: CurrencyReceiveWireframeContext,
        networksService: NetworksService
    ) {
        self.parent = parent
        self.context = context
        self.networksService = networksService
    }
}

extension DefaultCurrencyReceiveWireframe: CurrencyReceiveWireframe {
    
    func present() {
        let vc = wireUp()
        parent?.show(vc, sender: self)
    }
    
    func navigate(destination_______ destination: CurrencyReceiveWireframeDestination) {
        if (destination as? CurrencyReceiveWireframeDestination.Dismiss) != nil {
            vc?.popOrDismiss()
        }
    }
}

private extension DefaultCurrencyReceiveWireframe {
    
    func wireUp() -> UIViewController {
        let interactor = DefaultCurrencyReceiveInteractor(networksService: networksService)
        let vc: CurrencyReceiveViewController = UIStoryboard(.currencyReceive).instantiate()
        let presenter = DefaultCurrencyReceivePresenter(
            view: WeakRef(referred: vc),
            wireframe: self,
            interactor: interactor,
            context: context
        )
        vc.presenter = presenter
        self.vc = vc
        return vc
    }
}

File: androidApp/src/main/java/com/sonsofcrypto/web3wallet/android/modules/compose/currencyreceive/CurrencyReceiveWireframe.kt
package com.sonsofcrypto.web3wallet.android.modules.compose.currencyreceive

import androidx.fragment.app.Fragment
import com.sonsofcrypto.web3lib.services.networks.NetworksService
import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3wallet.android.common.extensions.navigationFragment
import com.sonsofcrypto.web3wallet.android.common.ui.navigationFragment
import com.sonsofcrypto.web3walletcore.modules.currencyReceive.*
import com.sonsofcrypto.web3walletcore.modules.currencyReceive.CurrencyReceiveWireframeDestination.Back
import com.sonsofcrypto.web3walletcore.modules.currencyReceive.CurrencyReceiveWireframeDestination.Dismiss

class DefaultCurrencyReceiveWireframe(
    private val parent: WeakRef<Fragment>?,
    private val context: CurrencyReceiveWireframeContext,
    private val networksService: NetworksService,
): CurrencyReceiveWireframe {

    override fun present() {
        val fragment = wireUp()
        parent?.navigationFragment?.push(fragment, animated = true)
    }

    override fun navigate(destination: CurrencyReceiveWireframeDestination) {
        when (destination) {
            Back -> { parent?.navigationFragment?.popOrDismiss() }
            Dismiss -> { parent?.navigationFragment?.dismiss() }
        }
    }

    private fun wireUp(): Fragment {
        val view = CurrencyReceiveFragment()
        val interactor = DefaultCurrencyReceiveInteractor(
            networksService
        )
        val presenter = DefaultCurrencyReceivePresenter(
            WeakRef(view),
            this,
            interactor,
            context
        )
        view.presenter = presenter
        return view
    }
}